A79-non-per-call-metrics-architecture.md

```diff
@@ -0,0 +1,405 @@
+       : target_(target), authority_(authority) {}
+
+   absl::string_view target() const { return target_; }
+   absl::string_view authority() const { return authority_; }
+
+  private:
+   absl::string_view target_;
+   absl::string_view authority_;
+ };
+
+ virtual bool IsEnabledForChannel(const ChannelScope& scope);
+ virtual bool IsEnabledForServer(ChannelArgs& args);
+
+ virtual void AddCounter(
+     GlobalInstrumentsRegistry::GlobalUInt64CounterHandle handle,
+     uint64_t value, absl::Span<const absl::string_view> label_values,
+     absl::Span<const absl::string_view> optional_values) = 0;
+ virtual void RecordHistogram(
+     GlobalInstrumentsRegistry::GlobalDoubleHistogramHandle handle,
+     double value, absl::Span<const absl::string_view> label_values,
+     absl::Span<const absl::string_view> optional_values) = 0;
+};
+
+class GlobalStatsPluginRegistry {
+public:
+ class StatsPluginGroup {
+  public:
+   // Use the stats plugin group object to record metrics
+   template <typename T>
+   void CounterAdd(GlobalUInt64CounterHandle handle, T value,
+                   RefCountedPtr<LabelValueSet> label_values,
+                   RefCountedPtr<LabelValueSet> optional_values);
+   template <typename T>
+   void HistogramRecord(GlobalDoubleHistogramHandle handle, T value,
+                        RefCountedPtr<LabelValueSet> label_values,
+                        RefCountedPtr<LabelValueSet> optional_values);
+
+   // Use the stats plugin to get a representation of label values that can be
+   // saved for multiple uses later.
+   RefCountedPtr<LabelValueSet> MakeLabelValueSet(
+       absl::Span<absl::string_view> label_values);
+ };
+
+ void RegisterStatsPlugin(std::shared_ptr<StatsPlugin> plugin);
+
+ // Invoked to get a stats plugin list for a specified scope.
+ static StatsPluginGroup GetStatsPluginsForChannel(ChannelScope scope);
+
+private:
+ std::vector<std::shared_ptr<StatsPlugin>> plugins_;
+};
+```
+
+##### Java
+
+(To be filled)
+
+##### Go
+
+(To be filled)
+
+### Metric Instrument Naming Conventions
+
+gRPC metrics follow the same general
+[guidelines](https://opentelemetry.io/docs/specs/semconv/general/metrics/#general-guidelines)
+as OpenTelemetry. The following additional guidelines apply -
+
+*   Metrics defined by gRPC should go under the `grpc.` namespace.
+*   Third-party plugins to gRPC should NOT use the `grpc.` namespace.
+*   Newer metrics should follow the same style as existing metrics. * Metrics
+    from load-balancing policies should be nested under `grpc.lb.`, e.g.
+    `grpc.lb.pick_first.`.
+*   Metrics from resolvers should be nested under `grpc.resolver.`, e.g.
+    `grpc.resolver.dns.`.
+*   Metrics from transports should be nested under `grpc.transport.`, e.g.,
+    `grpc.transport.http2.`
+
+### Metric Stability Guidelines and Guarantees
+
+From a stability perspective, metrics exported by gRPC will be in one of the
+following categories -
+
+*   On by default
```

Comment:
I think these options would be a bit clearer if presented as a flat list:

- Experimental, off by default
- Stable but off by default
- Stable and on by default

Reply:
done